pkg/prom/cluster/node.go

```diff
@@ -0,0 +1,327 @@
+	"github.com/go-kit/kit/log/level"
+	"github.com/google/go-cmp/cmp"
+	pb "github.com/grafana/agent/pkg/agentproto"
+	"github.com/grafana/agent/pkg/prom/ha/client"
+	"github.com/grafana/agent/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/user"
+)
+
+const (
+	// agentKey is the key used for storing the hash ring.
+	agentKey = "agent"
+)
+
+var backoffConfig = cortex_util.BackoffConfig{
+	MinBackoff: time.Second,
+	MaxBackoff: 2 * time.Minute,
+	MaxRetries: 10,
+}
+
+// node manages membership within a ring. when a node joins or leaves the ring,
+// it will inform other nodes to reshard their workloads. After a node joins
+// the ring, it will inform the local service to reshard.
+type node struct {
+	log log.Logger
+	reg *util.Unregisterer
+	srv pb.ScrapingServiceServer
+
+	mut  sync.RWMutex
+	cfg  Config
+	ring *ring.Ring
+	lc   *ring.Lifecycler
+
+	exited bool
+	reload chan struct{}
+}
+
+// newNode creates a new node and registers it to the ring.
+func newNode(reg prometheus.Registerer, log log.Logger, cfg Config, s pb.ScrapingServiceServer) (*node, error) {
+	n := &node{
+		reg: util.WrapWithUnregisterer(reg),
+		srv: s,
+		log: log,
+
+		reload: make(chan struct{}, 1),
+	}
+	if err := n.ApplyConfig(cfg); err != nil {
+		return nil, err
+	}
+	go n.run()
+	return n, nil
+}
+
+func (n *node) ApplyConfig(cfg Config) error {
+	n.mut.Lock()
+	defer n.mut.Unlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	// Detect if the config changed.
+	if cmp.Equal(n.cfg, cfg) {
+		return nil
+	}
+
+	if n.exited {
+		return fmt.Errorf("node already exited")
+	}
+
+	level.Info(n.log).Log("msg", "applying config")
+
+	// Shut down old components before re-creating the updated ones.
+	n.reg.UnregisterAll()
+
+	if n.lc != nil {
+		// Note that this will call performClusterReshard and will block until it
+		// completes.
+		err := services.StopAndAwaitTerminated(ctx, n.lc)
+		if err != nil {
+			return fmt.Errorf("failed to stop lifecycler: %w", err)
+		}
+	}
+
+	if n.ring != nil {
+		err := services.StopAndAwaitTerminated(ctx, n.ring)
+		if err != nil {
+			return fmt.Errorf("failed to stop ring: %w", err)
+		}
+	}
+
+	if !cfg.Enabled {
+		n.cfg = cfg
+		return nil
+	}
+
+	r, err := newRing(cfg.Lifecycler.RingConfig, "agent_viewer", agentKey, n.reg)
+	if err != nil {
+		return fmt.Errorf("failed to create ring: %w", err)
+	}
+	if err := n.reg.Register(r); err != nil {
+		return fmt.Errorf("failed to register ring metrics: %w", err)
+	}
+	if err := services.StartAndAwaitRunning(context.Background(), r); err != nil {
+		return fmt.Errorf("failed to start ring: %w", err)
+	}
+	n.ring = r
+
+	lc, err := ring.NewLifecycler(cfg.Lifecycler, n, "agent", agentKey, false, n.reg)
+	if err != nil {
+		return fmt.Errorf("failed to create lifecycler: %w", err)
+	}
+	if err := services.StartAndAwaitRunning(context.Background(), lc); err != nil {
+		r.StopAsync()
```

Comment:
Should this block?

Reply:
I don't think it would matter much here, other than it being a little bit slower, but I'll make it blocking to be safe.